### PR TITLE
Implement support for total counter on relationship fields

### PR DESCRIPTION
**Pull request by u12206050**
related-count

Had to find a way to include the display even though the field itself didn't have a value, therefore adding `item`
Also, had to skip prefixing of the function field which resulted in `relation.count(key)` which is obviously wrong.

### CODE REVIEW

**Comment by benhaynes** [comment]
Love it! Thank you for the PR!! ❤️

File: app/src/views/private/components/render-display/render-display.vue [hunk 1]
```diff
@@ -8,6 +8,7 @@
 		:interface="interface"
 		:interface-options="interfaceOptions"
 		:value="value"
+		:item="item"
```
[the remark this answers is not in the document]
Comment (u12206050) [reply app/src/views/private/components/render-display/render-display.vue]: Fixed by renaming to rootItem

**Comment by u12206050** [comment]
@rijkvanzanten I know this question is frowned upon, but :) Any timeline when this might be released?

**Comment by benhaynes** [comment]
Nothing official, but we're trying to get as many PRs reviewed and merged as possible. That said, we've been focusing on ones that resolve issues so far... but hopefully someone on the @directus/team can give it a lok!

**Review by rijkvanzanten — CHANGES_REQUESTED** [review]

**Comment by u12206050** [comment]
Ah cool I didn't even know about the fields property. How does it work?

**Comment by rijkvanzanten** [comment]
> Ah cool I didn't even know about the fields property. How does it work?
> […](#)

Instead of setting an array of fields, it can be a function get gets some context around the current configured field that's called, so you can do something like:

```js
fields: () => {
  return ['id'];
}
```

or in more practical sense:

```js
fields: (options, { field }) => [`count(${field.field})`]
```

File: app/src/displays/related-count/index.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+				},
+			},
+			schema: {
+				default_value: '$t:items',
+			},
+		},
+	],
+	types: ['alias'],
+	localTypes: ['m2m', 'o2m', 'translations', 'm2a', 'file', 'files'],
+	fields: ['count()'],
```
[the remark this answers is not in the document]
Comment (u12206050) [reply app/src/displays/related-count/index.ts:45]: That is why I had to override the adjust-fields-for-displays

[the remark this answers is not in the document]
Comment (u12206050) [reply app/src/displays/related-count/index.ts:45]: That is why I had to override the adjust-fields-for-displays

**Comment by u12206050** [comment]
> > Ah cool I didn't even know about the fields property. How does it work?
> > […](#)
> 
> Instead of setting an array of fields, it can be a function get gets some context around the current configured field that's called, so you can do something like:
> 
> ```js
> fields: () => {
>   return ['id'];
> }
> ```
> 
> or in more practical sense:
> 
> ```js
> fields: (options, { field }) => [`count(${field.field})`]
> ```

So I've tried this, seems like I already made use of is, and had to override `adjust-fields-for-display` otherwise the field just gets prepended making the query invalid. ie. `relationship.count()` it needs to be `count(relationship)`

**Comment by u12206050** [comment]
@azrikahar Would be nice to get this in the next release. Small PR but very nice functionality :)

**Comment by benhaynes** [comment]
I'm so surprised we don't have this "count" display yet. I defer to the engineers on if the PR is ready... but here are a few quick thoughts.

-  We should use the existing "Singular Unit" and "Plural Unit" settings within collection translations. @rijkvanzanten thoughts on adding these two fields to the main collection page too? They'd be more visible and easier to configure when not multilingual. Also, when _not_ configured let's ensure "Item" and "Items" is used as a fall back.
- This display should mention/show that the singular/plural names configured above are used (perhaps as a notice or placeholders) and perhaps offer overrides for each.
- Could we broaden this to work with JSON, arrays and CSVs too?
- Many of the formatting options available within "Format Value" display would be nice (which would always be true)... @rijkvanzanten I wonder if we should make those global for the field so they work with every display... or should we merge this display into that... or should we add those options to this display? 🤔

**Comment by azrikahar** [comment]
> @azrikahar Would be nice to get this in the next release. Small PR but very nice functionality :)

Granted I'm not opposed to it per se, I believe this is currently not slated for the next release based on our internal tracker (my understanding would be fixes takes precedence over new features as of now), so I'll have to defer to @rijkvanzanten for this. 👍

**Comment by rijkvanzanten** [comment]
> We should use the existing "Singular Unit" and "Plural Unit" settings within collection translations. @rijkvanzanten thoughts on adding these two fields to the main collection page too? They'd be more visible and easier to configure when not multilingual. Also, when not configured let's ensure "Item" and "Items" is used as a fall back.

Yeah, and the collection translations setup needs to use the updated translations inputs, rather than the older repeater style config it's still using currently 👍🏻 

> Many of the formatting options available within "Format Value" display would be nice (which would always be true)... @rijkvanzanten I wonder if we should make those global for the field so they work with every display... or should we merge this display into that... or should we add those options to this display? 

While some displays could benefit from some of those shared options, I definitely don't think it applies to most of them. Lets not convolute the field configuration setup by trying to configure display information on both the field _and_ display level 👍🏻

**Comment by benhaynes** [comment]
> Yeah, and the collection translations setup needs to use the updated translations inputs, rather than the older repeater style config it's still using currently 👍🏻

Beautiful!! I didn't think about that, but it really simplifies this and solves both those issues.

> While some displays could benefit from some of those shared options, I definitely don't think it applies to most of them. Lets not convolute the field configuration setup by trying to configure display information on both the field and display level 👍🏻

Hmm, ok! So then we'd want to recreate the "Conditional Styles" option from within the "Formatted Value" Display. We should at least be consistent with that GUI across different Displays.

**Comment by u12206050** [comment]
I'm a bit unsure of changes to actually implement or are the things @benhaynes and @rijkvanzanten are talking about something that is done while merging in?

**Comment by benhaynes** [comment]
> I'm a bit unsure of changes to actually implement or are the things @benhaynes and @rijkvanzanten are talking about something that is done while merging in?

We have "Conditional Formatting" on other Displays which allows you to change the style (color, etc) of the display output based on the value (in this case, a count). It's useful here too, as you  might need to show `0 Items` in **red**, or when you have `10+ items` in **green**, etc.

Should be easy to copy from this other display. The main applicable style options here are color and icon.

https://github.com/directus/directus/blob/main/app/src/displays/formatted-value/index.ts#L158

**Comment by u12206050** [comment]
Done. Hope it is ok that we always show a value even if it is 0. Logically it just makes sense in most scenarios.

**Comment by u12206050** [comment]
Sorry was struggling getting eslint to work on my local machine so had to fix the errors manually.
Should be all good now.

**Comment by u12206050** [comment]
@benhaynes @rijkvanzanten I have cleaned this PR up as well, reusing the formatted field display to actually display the options instead of duplicating it.

o2mcounter

**Comment by u12206050** [comment]
@benhaynes would the latest aggregation query and functions on fields be able to solve this if just exposed as being selectable in columns or is this still the way we would want it?

**Comment by benhaynes** [comment]
> @benhaynes would the latest aggregation query and functions on fields be able to solve this if just exposed as being selectable in columns or is this still the way we would want it?

Oooh, I like that! I defer to @rijkvanzanten on the technical feasibility of this (how we track fields versus aliases)... but I like the idea, just like how it's displayed in the filters.

**Comment by rijkvanzanten** [comment]
Aggregations run on a set of data, not single rows, so this would be exclusive to function support 👍🏻

**Comment by u12206050** [comment]
> 

Ok, nice. So do we close this PR in favour of using functions as columns instead or? @rijkvanzanten

**Comment by rijkvanzanten** [comment]
Yeah lets do it 👍🏻

**Pull request closed without merging**